---
title: Promise
date: 2021-04-21 09:46:23
permalink: /pages/d57563/
categories:
  - 前端
tags:
  - JavaScript
  - 异步编程
---
## 前言
Promise 是 ES6 引入的异步编程方案，能有效解决回调地狱问题。简单来说，Promise 是一个对象，保存着未来才会结束的事件，并提供统一的 API 供使用者处理异步操作。

## 创建
### new Promise()
Promise 构造函数接受函数作为参数，该函数参数为 resolve 和 reject。通过 new Promise() 来创建 promise 对象最为常见
```js
const promise = new Promise(function (resolve, reject) {
  // 一些操作后会调用 resolve 或 reject
  if (/* 一些判断 */) {
    resolve(someValue) // fulfilled 状态，并返回值
  } else {
    reject(/* 失败原因 */) // rejected 状态
  }
});
```
新生成的 promise 对象会处于 pending(进行中) 状态，后面根据执行结果进入 fulfilled（已成功）或 reject（已失败）状态

### Promise.resolve()
```js
Promise.resolve(1).then(function (value) {
  console.log(value); // 1
});
```
这种形式的 promise 对象已经是 resolve 状态，并将参数传递到 then 所指定的 onFulfilled 函数

### Promise.reject()
```js
Promise.reject(new Error('进入 rejected 状态'));
```

## Promise 解决回调地狱
先看一下回调地狱
```js
let fs = require('fs')
fs.readFile('./a.txt','utf8', function (err, data) {
  fs.readFile(data, 'utf8', function (err, data) {
    fs.readFile(data, 'utf8', function (err, data) {
      console.log(data)
    })
  })
})
```
缺点：
- 代码嵌套多层，不直观
- 每个人物需要分贝处理成功和失败两种可能性

使用 Promise 来看看
```js
let fs = require('fs')
function read (url) {
  return new Promise((resolve,reject) => {
    fs.readFile(url, 'utf8', function (err, data) {
      err && reject(err)
      resolve(data)
    })
  })
}
read('./a.txt').then(data => {
  return read(data)
}).then(data => {
  return read(data)
}).then(data => {
  console.log(data)
})
```
看起来就简洁了不少，很好地解决了回调地狱，它是怎么解决的呢？

首先解决嵌套多层，Promise 采用回调函数延迟绑定和

### 回调函数延迟绑定
从上面例子中看出回调函数是 read.then() 中传递的，而不是在声明 read 时传入的，这样的延迟传入，

## API
### Promise.then()

### Promise.catch()

### Promise.all()

### Promise.race()

### Promise.finally()

### Promise.allSettled()

### Promise.any()